### ISSUE TITLE
Some refactoring of SoundExpressions

### ISSUE THREAD

**Issue by pelikhan** [lancaster-university/codal-microbit-v2#4]
Suggestion to the current implementation:

- [ ] use a binary format that can be mapped to a C struct
- [ ] removing all parsing code, use of ManagedString
- [ ] convert existing sounds to binary format

Once this is done, we can lift the sound definitions to MakeCode resources so that the simulator can eventually play them.

**Comment by finneyj** [comment]
The underlying SoundEffect used by the Synth is a binary format that I think would suit if we get a nice JS editor to create sound effects.

I have a ticket to write a frequency slide effect that would reduce most of the current "SoundEmojis" to one or two binary SoundEffect frames. Once we have that, would the SoundEffect structure (and friends) suit @pelikhan ?

https://github.com/lancaster-university/codal-microbit-v2/blob/master/inc/SoundEmojiSynthesizer.h#L94

**Comment by pelikhan** [comment]
Yes it looks better. Eventually we should be able to build an editor for it. What is the relation of soundeffect to the previous CODAL sound pipeline?

________________________________

**Comment by finneyj** [comment]
Great - thanks @pelikhan. I thought you might like to write an editor for this in due course. That would give a great experience I think. I'd like to upstream this audio pipeline once we've stabilized it here I think (and replace the Mixer/Synthesizer in codal-core)... so this could also be useful for a future version of arcade etc.

The SoundExpressions class is just a thin wrapper that creates a series of SoundEffect structures from a String serialization. @microbit-matt-hillsdon  wrote that for us to interface with a JS design tool that some sound engineers have been working with for @jaustin.

Audio output pipeline looks like this:

![image](https://user-images.githubusercontent.com/11539567/97201610-8c42cf80-17aa-11eb-8133-119aa08c9f65.png)

**Comment by jaustin** [comment]
Definitely up for a nice editor for these synth sounds (we did a tool for the sound designer to use and it was really fun) . I think a binary serialisation makes a lot of sense, but removing all parsing code from the DAL is not a good answer because it means we don't have consistency across editors for sound parsing - one of the key reasons CODAL exists. Would be happy for DAL to assume correctly-formed binary serialisation as long as that is defined in CODAL. 

Whatever the serialisation format, it needs to be possible for a cool sound made to sound the same in any editor building on CODAL. This is part of the promise for people using CODAL is that they get al micro:bit features, and this is a key feature The Foundation has been designing for v2.

We plan to continue to evolve and add to the sounds, and propogating them out from the DAL is the best way to do this, so having a set of built-in sounds remains important.

@dpgeorge at the moment you're just using the built-in sounds, right? Any requirements around serialisation format from your end?

**Comment by dpgeorge** [comment]
> at the moment you're just using the built-in sounds, right? Any requirements around serialisation format from your end?

At the moment MicroPython exposes them like `audio.play(Sound.HAPPY)` and only supports built-in sounds.  There are therefore no requirements for serialisation.
